### ISSUE TITLE
Pausing autoscaling

### ISSUE THREAD

**Issue by kanaluM-crowdstrike** [jthomperoo/custom-pod-autoscaler-operator#105]
__Is your feature request related to a problem? Please describe.__  
It would be helpful if there was a way to easily pause and unpause autoscaling (ex. for cluster maintenance).

__Describe the solution you'd like__  
*To avoid confusion, the pod(s) created by this `custom-pod-autoscaler-operator` repo will be referred to as the `operator` and the pod(s) created by the [other](https://github.com/jthomperoo/custom-pod-autoscaler) `custom-pod-autoscaler` repo will be referred to as the `autoscaler`. The pods that are being scaled by the autoscaler (ex. a microservice) will be referred to as the `target`.

The proposed solution is to use a specific annotation (key/value pair) in the `autoscaler`'s configuration file (ex. Helm chart or `.yaml` file). When the annotation is present, it will pause autoscaling. When it is not present, autoscaling will run as normal. See __Additional context__ for more information on why this approach is good (tldr: copy [KEDA](https://keda.sh/docs/2.11/concepts/scaling-deployments/#pause-autoscaling)).

The annotation's _key_ would be a special string that the `operator` will look for. The _value_ would be the number of replicas you desire for the `target` while autoscaling is paused (let's call this number the `pausedReplicasCount`). For example, the presence of the annotation in the example below would pause autoscaling and set the replica count of the `target` to 42.
```
apiVersion: custompodautoscaler.com/v1
kind: CustomPodAutoscaler
metadata:
  name: custom-autoscaler
  annotations:
    "v1.custompodautoscaler.com/paused-replicas": "42"
spec:
  template:
```
The proposed implementation for this would mainly involve a change to the function
```
func (r *CustomPodAutoscalerReconciler) Reconcile(context context.Context, req ctrl.Request)
```
found in `controllers/custompodautoscaler_controller.go`. 

The proposed implementation is the following:
- The `operator` detects the pause annotation on an `autoscaler`
- The `operator` deletes the `autoscaler`
- The `operator` updates `target` to have `pausedReplicasCount` replicas

If the annotation is not detected, the `operator` will reconcile the `autoscaler` as usual. This includes creating an `autoscaler` if one does not exist

__Describe alternatives you've considered__  
The proposed implementation is to have the `operator` do everything by itself.

One alternative is to have the `autoscaler` do everything by itself. This would be difficult because it is not trivial to have the `autoscaler` query for metadata information about itself (via something like the downward API). It is also not feasible for the `autoscaler` to delete itself while paused (and then come back when unpaused). It is possible to let the `autoscaler` remain intact and simply disable its autoscaling functionality while paused, but this would be more trouble than it's worth. Seeing that the `autoscaler` has been deleted will be a clearer sign to cluster admins that things are going as expected.

The second and more promising alternative is to have the `operator` and `autoscaler` share the work. The workflow would look something like this:
- The `operator` detects the annotation and sends the `pausedReplicasCount` to the `autoscaler`
- The `autoscaler` updates `target` to have `pausedReplicasCount` replicas
- The `autoscaler` informs the `operator` when this change has been successfully made
- The `operator` deletes the `autoscaler`

The main _advantage_ of this alternative approach is that the `autoscaler` already has the necessary infrastructure to scale pods while the `operator` does not. The `operator` also by default does not have RBAC permission to scale pods (this would require developers to allow that in the Helm chart or other configuration file)

The main _disadvantage_ of this approach is that it would be quite complicated to implement. It would require changes to both the `custom-pod-autoscaler-operator` and `custom-pod-autoscaler` repos. This would make it much more difficult to review and understand the changes. 

Overall it makes more sense to go for the simpler approach and have the `operator` do everything. It is not too difficult to give the `operator` some basic scaling capabilities (it is more or less a copy-and-paste from the `custom-pod-autoscaler` repo). It is also not very costly for the `operator` to have another client for scaling. These changes could be made with just a few additional client dependencies that the `custom-pod-autoscaler` repo also uses. Hopefully these changes will be easy to review and understand.

__Additional context__  
KEDA recently implemented a [feature](https://keda.sh/docs/2.11/concepts/scaling-deployments/#pause-autoscaling) to pause and unpause HPA autoscaling via a metadata annotation. It makes sense for CPA to have similar functionality and implementation of the feature. I mostly remained faithful to the KEDA implementation for _checking_ for a pause annotation. However, the KEDA implementation for _setting_ the `target` replicas to `pausedReplicasCount` is rather complicated due to their architecture being much more complex. KEDA also just uses built-in Kubernetes HPA pods for autoscaling, while this project uses its own custom pods with source code in a separate repo. I think these proposed changes to the CPA operator will be a bit easier for other developers to understand while still keeping the operator somewhat lightweight.

**Comment by jthomperoo** [comment]
Thanks for the detailed write up and pull request, I'll look through all of this over the next few days and hopefully we can get it merged in :)

**Comment by jthomperoo** [comment]
Resolved by #106, available in release [`v1.4.0`](https://github.com/jthomperoo/custom-pod-autoscaler-operator/releases/tag/v1.4.0).

See the docs for this feature here:

https://github.com/jthomperoo/custom-pod-autoscaler-operator/blob/v1.4.0/USAGE.md#pausing-autoscaling
